src/LoadBalance/LB_GatherTree.cpp

```diff
@@ -767,3 +767,275 @@ LB_GlobalPatch* LB_GatherTree( LB_PatchCount& pc, int root ) {
+   }
+#  endif
+
+   return Patches[GID].corner[XYZ] + I * amr->scale[Patches[GID].level];
+} // LB_GlobalTree::Local2Global
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  LB_GlobalTree::FindRefinedCounterpart
+// Description :  Return GID of refined patch with maximum level MaxLv (default ) that global coordinates [X, Y, Z] belong to
+//
+//
+// Parameter   :  I   : x-index of patch GID
+//             :  J   : y-index of patch GID
+//             :  K   : z-index of patch GID
+//             : GID  : global ID of patch
+//             : MaxLv: maximum refinemen
+//
+// Return      :  GID of patch
+//-------------------------------------------------------------------------------------------------------
+long LB_GlobalTree::FindRefinedCounterpart(int X, int Y, int Z, long GID, int MaxLv)  const
+{
+
+
+#  ifdef GAMER_DEBUG
+// sanity check
+   if (GID == -1)
+   {
+      Aux_Error(ERROR_INFO, "GID  == -1!\n");
+   }
+   if (GID >= NPatch)
+   {
+      Aux_Error(ERROR_INFO, "GID (%ld) >= NPatch (%ld)!\n", GID, NPatch);
+   }
+#  endif
+
+   if ( Patches[GID].level > MaxLv )
+   {
+      return -1;
+   }
+
+
+// skip calculation if coordinates are not inside patch GID
+   if ( !IsInsidePatch(X, Y, Z, GID) )
+   {
+      return -1;
+   }
+
+   long FaGID  = GID;
+   long SonGID = Patches[FaGID].son;
+
+// traverse the tree up until leave nodes
+   while ( SonGID != -1 ) {
+
+//    exit loop if FaGID is on MaxLv
+      if ( Patches[SonGID].level > MaxLv )
+      {
+         break;
+      }
+
+//    loop over patches in patch block and check if cell {K, J, I} belongs to patch
+      for (int LocalPID = 0; LocalPID < 8; LocalPID++)
+      {
+
+         if ( IsInsidePatch(X, Y, Z, SonGID + LocalPID) )
+         {
+            FaGID   = SonGID + LocalPID;
+            SonGID  = Patches[FaGID].son;
+            break;
+
+         }
+#        ifdef GAMER_DEBUG
+         else if ( LocalPID == 7 )
+         {
+            Aux_Error(ERROR_INFO, "Global coordinates {%d, %d, %d} in father patch (GID = %ld, lv = %d), but not in any son patch (GID = %ld, lv = %d)!!\n",
+            X, Y, Z, FaGID, Patches[FaGID].level, SonGID, Patches[SonGID].level);
+         }
+#        endif
+      }
+   }
+
+#  ifdef GAMER_DEBUG
+// sanity check
+   if (FaGID == -1)
+   {
+      Aux_Error(ERROR_INFO, "FaGID == -1!!\n");
+   }
+
+// check whether GID is ancestor of FaGID
+// number of generations between GID and FaGID
+   int  NGenerations = Patches[FaGID].level - Patches[GID].level;
+
+// iterate back through ancestors
+   long AncestorGID = FaGID;
+   for (int i = 0; i < NGenerations; ++i)
+   {
+      AncestorGID = Patches[AncestorGID].father;
+   }
+
+   if (Patches[AncestorGID].level != Patches[GID].level)
+   {
+      Aux_Error(ERROR_INFO, "GID (%ld) and Ancestor (%ld) on different levels!!\n", GID, AncestorGID);
+   }
```

Comment:
We may remove it since the next check `if (AncestorGID != GID)` is more rigorous.